gen.go

```diff
@@ -1591,7 +1651,23 @@ func (t *{{ .Name}}) UnmarshalCBOR(r io.Reader) (err error) {
 		} else {
 			f.Name = "t." + f.Name
 		}
-		fmt.Fprintf(w, "\t// %s (%s) (%s)\n", f.Name, f.Type, f.Type.Kind())
+
+		if f.Type.String() == gti.GenericPlaceholder {
+			g.doTemplate(w, f, `
+				// {{ .Name }} (T)
+		    {
+					var value T
+					value = value.New()
+					if err := value.UnmarshalCBOR(cr); err != nil {
+						return xerrors.Errorf("failed to read field: %w", err)
+					}
```

Comment:
We should probably make a helper `NewFromCBOR` function.

Reply:
Also, this is moderately annoying because we only need `\.New()` if we have a pointer. This goes back to `**` land, but... what if we required the user to implement `UnmarshalCBOR`  on `**` if they want a pointer? That is:

```go
func (t **T) UnmarshalCBOR(r io.Reader) error {
	*t = new(T)
	return (*t).UnmarshalCBOR(r)
}
```

Reply:
Ah, wait, no... we can't do that. We'd need an additional constraint/generic parameter.

Reply:
Yeah, this is the bit I'm struggling with the most, we also can't check for pointer types at gen time and use them to handle the nullable case, `T` is never a pointer as far as `reflect` is concerned.

I think for now, you're just going to have to be bound to implementing `UnmarshalCBOR` as a pointer receiver and therefore your type will just need to be a pointer for any of this to work. And, you'll have to handle the nullable case in your own serialization code.

Smarter can come later if someone wants to come up with a better approach.

Reply:
I'm wondering if we should change the interface for generic types (possibly providing an adapter, not sure if we can roll that in go...)?

```go
type GenericCBOR[T] interface {
	FromCBOR(r io.Reader) *T
	ToCBOR(w io.Writer)
}
```

This interface would take the receiver by-value (and I'm hoping it won't have too much overhead from copying).